File: typescript/creational/abstract_factory/furniture/factory.ts
import {
  Chair, ArtDecoChair, VictorianChair, ModernChair,
  Table, ArtDecoTable, VictorianTable, ModernTable,
  Sofa, ArtDecoSofa, VictorianSofa, ModernSofa
} from './product'

export interface Factory {
  createChair(): Chair,
  createTable(): Table,
  createSofa(): Sofa
}

export class ArtDeco implements Factory {
  public createChair(): Chair {
    return new ArtDecoChair()
  }
  public createTable(): Table {
    return new ArtDecoTable()
  }
  public createSofa(): Sofa {
    return new ArtDecoSofa()
  }
}

export class Victorian implements Factory {
  public createChair(): Chair {
    return new VictorianChair()
  }
  public createTable(): Table {
    return new VictorianTable()
  }
  public createSofa(): Sofa {
    return new VictorianSofa()
  }
}

export class Modern implements Factory {
  public createChair(): Chair {
    return new ModernChair()
  }
  public createTable(): Table {
    return new ModernTable()
  }
  public createSofa(): Sofa {
    return new ModernSofa()
  }
}

File: dart/creational/abstract_factory/furniture/bin/factory.dart
import 'product.dart';

abstract class Factory {
  Chair createChair();
  Table createTable();
  Sofa createSofa();
}

class ArtDeco implements Factory {
  @override
  Chair createChair() {
    return ArtDecoChair();
  }

  @override
  Table createTable() {
    return ArtDecoTable();
  }

  @override
  Sofa createSofa() {
    return ArtDecoSofa();
  }
}

class Victorian implements Factory {
  @override
  Chair createChair() {
    return VictorianChair();
  }

  @override
  Table createTable() {
    return VictorianTable();
  }

  @override
  Sofa createSofa() {
    return VictorianSofa();
  }
}

class Modern implements Factory {
  @override
  Chair createChair() {
    return ModernChair();
  }

  @override
  Table createTable() {
    return ModernTable();
  }

  @override
  Sofa createSofa() {
    return ModernSofa();
  }
}
